compiler/frontend/src/print.rs

```diff
@@ -0,0 +1,304 @@
+            } else {
+                "…".to_string()
+            }
+        }
+        PrintValue::Function => {
+            // - `{ … }`
+            // - `…`
+
+            if max_length.fits(5) { "{ … }" } else { "…" }.to_string()
+        }
+        PrintValue::List(list) => {
+            // - all items: `(Foo, Bar, Baz)`
+            // - some items: `(Foo, Bar, + 2 more)`
+            // - no items shown: `(list of 2 items)`
+            // - `…`
+
+            if !max_length.fits(3) {
+                return Some("…".to_string());
+            }
+
+            if list.is_empty() {
+                return Some("(,)".to_string());
+            }
+
+            let list_len = list.len();
+            let mut items = vec![];
+            let mut total_item_length = 0;
+            for item in list {
+                let item = print(item, Precedence::Low, MaxLength::Unlimited, visitor)?;
+                total_item_length += item.len();
+                items.push(item);
+
+                // surrounding parentheses, items, and for each item comma + space
+                if !max_length.fits(2 + total_item_length + items.len() * 2) {
+                    break;
+                }
+            }
+            if items.len() == list_len && max_length.fits(total_item_length + items.len() * 2) {
+                return Some(format!("({})", items.into_iter().join(", ")));
+            }
+
+            // Not all items fit. Try to remove the back ones, showing "+ X more" instead.
+            while let Some(popped) = items.pop() {
+                total_item_length -= popped.len();
+                let extra_text = format!("+ {} more", list_len - items.len());
+                if max_length.fits(total_item_length + items.len() * 2 + extra_text.len()) {
+                    return Some(format!(
+                        "({}, {})",
+                        items.into_iter().join(", "),
+                        extra_text
+                    ));
+                }
+            }
+
+            let summary = format!("(list of {} items)", list_len);
+            if max_length.fits(summary.len()) {
+                summary
+            } else {
+                "…".to_string()
+            }
+        }
+        PrintValue::Struct(entries) => {
+            // - all entries: `[Foo: Bar, Baz: 2]`
+            // - all keys, some values: `[Foo: Bar, Baz: …, Quz: …]`
+            // - some keys: `[Foo: …, Bar: …, + 2 more]`
+            // - no items shown: `[struct with 2 entries]`
+            // - `…`
+
+            if !max_length.fits(2) {
+                return Some("…".to_string());
+            }
+
+            if entries.is_empty() {
+                return Some("[]".to_string());
+            }
+
+            let num_entries = entries.len();
+            let mut keys = vec![];
+            let mut values = vec![];
+            for (key, value) in entries {
+                keys.push(key);
+                values.push(value);
+            }
```

Comment:
I don't think these are even necessary, you can just iterate over the original `entries`